lib/util/pre_receive_util.js

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2016, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+/**
+ * This module contains the pre-recieve hook for a git-meta server.  Only
+ the meta-repo should use this hook.
+ */
+
+const NodeGit = require("nodegit");
+const GitUtil = require("./git_util");
+const async = require("generator-async");
+const split = require("split");
+const co = require("co");
+const path = require("path");
+const assert  = require("chai").assert;
+
+const NOTES_REF = "refs/notes/git-meta/subrepo-check";
+
+function flatmap(arr, fn) {
+    return arr.map(fn).filter(function (v) { return v; });
+}
+
+class PreReceiveCheck {
+
+    *urlToLocalPath(repo, url) {
+        const config = yield repo.config();
+        const subrepoUrlBase =
+            yield config.getStringBuf("gitmeta.subrepourlbase");
+        const subrepoRootPath =
+            yield config.getStringBuf("gitmeta.subreporootpath");
+        assert(url.startsWith(subrepoUrlBase));
+        const remotePath = url.slice(subrepoUrlBase.length);
+        const localPath = path.join(subrepoRootPath, remotePath);
+        if (localPath[0] === "/") {
+            return localPath;
+        } else {
+            return path.normalize(path.join(repo.path(), localPath));
+        }
+    }
+
+    getSyntheticBranchForCommit(commit) {
+        return "refs/git-meta/synthetic/meta-" + commit.id();
```

Comment:
This doesn't seem to agree with README.md, which says `/refs/heads/meta-$METASHA`